File: Matrix Convolution/matrix_convolution.c
#include <stdio.h> 
#include <stdlib.h>
#include <omp.h>
#include <math.h>
#include <time.h>
#define FILTER_SIZE 3

void convolution(int** input, int filter[FILTER_SIZE][FILTER_SIZE], int** output, int size, int padding);
void convolution_parallel(int** input, int filter[FILTER_SIZE][FILTER_SIZE], int** output, int size, int padding);
void freeMatrix(int** matrix, int size);
int** createMatrix(int size);
void printMatrix(int** matrix, int size);


int main(int argc, char* argv[]) {

	clock_t start, end, start_p, end_p;
	double execution_time, execution_time_p;
	
	/* Check if matrix size is provided */
	if (argc < 2) {
		fprintf(stderr, "Please insert matrix size.");
		return 1;
	}

	/* Get matrix size from command line */
	int matrix_size = strtol(argv[1], NULL, 10);

	/* Allocate memroy for intput matrix*/
	int** matrix = createMatrix(matrix_size);
	/* Allocate memory for output matrix */
	int output_size = matrix_size - FILTER_SIZE + 1;
	int** output_matrix = createMatrix(output_size);

	/* Fill the matrix with random numbers */
	srand(time(NULL));
	for (int rol = 0; rol < matrix_size; rol++)
	{
		for (int col = 0; col < matrix_size; col++) 
		{
			matrix[rol][col] = rand() % 10;
		}
	}	

	/* Print the matrix */
	//printf("Input Matrix:\n");
	//printMatrix(matrix, matrix_size);

	/* Create the filter matrix*/
	int filter[FILTER_SIZE][FILTER_SIZE] = {
		{0, -1, 0},
		{-1, 5, -1},
		{0, -1, 0}
	};

	/* Apply matrix convolution */
	int padding = 1;
	start = clock();
	convolution(matrix, filter, output_matrix, matrix_size, padding);
	end = clock();
	execution_time = ((double) end - start) / CLOCKS_PER_SEC;

	start_p = clock();
	convolution_parallel(matrix, filter, output_matrix, matrix_size, padding);
	end_p = clock();
	execution_time_p = ((double) end_p - start_p) / CLOCKS_PER_SEC;
        
	/* Print the output matrix */
	//printf("Output Matrix:\n");
	/* printMatrix(output_matrix, output_size); */

	/* Print the execution time */
	printf("Execution time: %f seconds\n", execution_time);
	printf("Execution time using parallelism: %f seconds\n", execution_time_p);

        /* Free the matrix */
	freeMatrix(matrix, matrix_size);
	freeMatrix(output_matrix, output_size);

}

void convolution (int** input, int filter[FILTER_SIZE][FILTER_SIZE], int** output, int input_size, int padding)
{
	for (int i = padding; i < input_size - padding; i++) {
		for (int j = padding; j < input_size - padding; j++) {
			int sum = 0;

			// Apply the filter matrix over the input matrix region
			for (int fi = -1; fi < FILTER_SIZE - 1; fi++) {
				for (int fj = -1; fj < FILTER_SIZE - 1; fj++) {
					sum += input[i + fi][j + fj] * filter[fi + 1][fj + 1];
				}
			}

			// Store the result in the output matrix
			output[i - padding][j - padding] = sum;
		}
	}
}

void convolution_parallel (int** input, int filter[FILTER_SIZE][FILTER_SIZE], int** output, int input_size, int padding)
{
        //omp_set_num_threads(4);
	int chunk_size = 256;
	printf("Parallel with chunk size: %d\n", chunk_size);
	#pragma omp parallel for schedule(static, 256)
        for (int i = padding; i < input_size - padding; i++) {
                for (int j = padding; j < input_size - padding; j++) {
                        int sum = 0;

                        // Apply the filter matrix over the input matrix region
                        for (int fi = -1; fi < FILTER_SIZE - 1; fi++) {
                                for (int fj = -1; fj < FILTER_SIZE - 1; fj++) {
                                        sum += input[i + fi][j + fj] * filter[fi + 1][fj + 1];
                                }
                        }

                        // Store the result in the output matrix
                        output[i - padding][j - padding] = sum;
                }
        }
}

void freeMatrix(int** matrix, int size)
{
	/* Free the matrix */
        for (int rol = 0; rol < size; rol++)
        {
                free(matrix[rol]);
        }
        free(matrix);
}

int** createMatrix(int size)
{
        int** matrix = malloc(size*sizeof(int*));
        for (int rol = 0; rol < size; rol++)
        {
                matrix[rol] = malloc(size*sizeof(int));
        }
	return matrix;
}

void printMatrix(int** matrix, int size)
{
        for (int rol = 0; rol < size; rol++)
        {
                for (int col = 0; col < size; col++)
                {
                        printf(" %d ", matrix[rol][col]);
                }
                printf("\n");
        }
}

File: Matrix Convolution/matrix_convolution.rs
use rand::Rng;
use rayon::prelude::*;
use std::time::Instant;

const FILTER_SIZE: usize = 3;

fn main() {
    // Get matrix size from command line
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("Please insert matrix size.");
        return;
    }

    let matrix_size: usize = args[1].parse().expect("Invalid matrix size");

    // Allocate memory for input matrix
    let mut input_matrix = create_matrix(matrix_size);
    let output_size = matrix_size - FILTER_SIZE + 1;
    let mut output_matrix = create_matrix(output_size);

    // Fill the matrix with random numbers
    let mut rng = rand::thread_rng();
    for row in input_matrix.iter_mut() {
        for elem in row.iter_mut() {
            *elem = rng.gen_range(0..10);
        }
    }

    //println!("Input Matrix:");
    //print_matrix(&input_matrix);
    
    // Define the filter matrix
    let filter: [[i32; FILTER_SIZE]; FILTER_SIZE] = [
        [0, -1, 0],
        [-1, 5, -1],
        [0, -1, 0],
    ];

    // Apply matrix convolution (single-threaded)
    let padding = 1;
    let start = Instant::now();
    convolution(&input_matrix, &filter, &mut output_matrix, matrix_size, padding);
    let duration = start.elapsed();
    println!("Execution time (single-threaded): {:?}", duration);

    //println!("Output Matrix (Sequential):");
    //print_matrix(&output_matrix);

    // Apply matrix convolution (parallelized)
    let start_p = Instant::now();
    convolution_parallel(&input_matrix, &filter, &mut output_matrix, matrix_size, padding);
    let duration_p = start_p.elapsed();
    println!("Execution time (parallelized): {:?}", duration_p);

    //println!("Output Matrix (Parallel):");
    //print_matrix(&output_matrix);
}

fn print_matrix(matrix: &Vec<Vec<i32>>) {
    for row in matrix {
        for &value in row {
            print!("{:3} ", value);
        }
        println!();
    }
    println!();
}

fn create_matrix(size: usize) -> Vec<Vec<i32>> {
    vec![vec![0; size]; size]
}

fn convolution(
    input: &Vec<Vec<i32>>,
    filter: &[[i32; FILTER_SIZE]; FILTER_SIZE],
    output: &mut Vec<Vec<i32>>,
    size: usize,
    padding: usize,
) {
    for i in padding..size - padding {
        for j in padding..size - padding {
            let mut sum = 0;
            for fi in 0..FILTER_SIZE {
                for fj in 0..FILTER_SIZE {
                    sum += input[i + fi - 1][j + fj - 1] * filter[fi][fj];
                }
            }
            output[i - padding][j - padding] = sum;
        }
    }
}

fn convolution_parallel(
    input: &Vec<Vec<i32>>,
    filter: &[[i32; FILTER_SIZE]; FILTER_SIZE],
    output: &mut Vec<Vec<i32>>,
    size: usize,
    padding: usize,
) {
    output.par_iter_mut().enumerate().for_each(|(i, row)| {
        if i >= padding && i < size - padding {
            for j in 0..row.len() {
                if j >= padding && j < size - padding {
                    let mut sum = 0;
                    for fi in 0..FILTER_SIZE {
                        for fj in 0..FILTER_SIZE {
                            sum += input[i + fi - 1][j + fj - 1] * filter[fi][fj];
                        }
                    }
                    row[j - padding] = sum;  // Adjust index to account for padding in the output
                }
            }
        }
    });
}

